ros/docker/launch.py

```diff
@@ -0,0 +1,119 @@
+#!/usr/bin/env python3
+
+import datetime
+import logging
+import os
+import re
+import signal
+import sys
+import telnetlib
+
+import vrnetlab
+
+def handle_SIGCHLD(signal, frame):
+    os.waitpid(-1, os.WNOHANG)
+
+def handle_SIGTERM(signal, frame):
+    sys.exit(0)
+
+signal.signal(signal.SIGINT, handle_SIGTERM)
+signal.signal(signal.SIGTERM, handle_SIGTERM)
+signal.signal(signal.SIGCHLD, handle_SIGCHLD)
+
+TRACE_LEVEL_NUM = 9
+logging.addLevelName(TRACE_LEVEL_NUM, "TRACE")
+def trace(self, message, *args, **kws):
+    # Yes, logger takes its '*args' as 'args'.
+    if self.isEnabledFor(TRACE_LEVEL_NUM):
+        self._log(TRACE_LEVEL_NUM, message, args, **kws)
+logging.Logger.trace = trace
+
+class ROS_vm(vrnetlab.VM):
+    def __init__(self, username, password):
+        for e in os.listdir("/"):
+            if re.search(".vmdk$", e):
+                disk_image = "/" + e
+        super(ROS_vm, self).__init__(username, password, disk_image=disk_image, ram=256)
+        self.qemu_args.extend(["-boot", "n"])
+        # Starting off with 8 nics
+        self.num_nics = 7
+
+    def bootstrap_spin(self):
+        """ This function should be called periodically to do work.
+        """
+
+        if self.spins > 300:
+            # too many spins with no result ->  give up
+            self.stop()
+            self.start()
+            return
+
+        (ridx, match, res) = self.tn.expect([b"MikroTik Login"], 1)
+        if match: # got a match!
+            if ridx == 0: # login
+                self.logger.debug("VM started")
+                self.tn = telnetlib.Telnet("127.0.0.1", 5000 + self.num)
+
+                # Login
+                self.wait_write("\r", None)
+                # Append +ct to username for the plain-text console version
```

Comment:
Actually, the documentation wording is weird. "+ct" disables funky colors and terminal capabilities auto-detect. It's only used for the bootstrap process. I use this mode when automating MikroTiks to prevent RouterOS from returning weird colour attributes.